### PR TITLE
Make Auto-GPT aware of its running cost

### PR DESCRIPTION
**Pull request by Vwing**
### Background
This pull request addresses issue #6, which aims to make Auto-GPT aware of its running cost. The changes centralize all OpenAI Chat Completion calls through a single method that counts tokens and adds to a dollar amount based on the model used. The total dollar amount spent is constantly displayed to the user and is shown to the AI as well.

The user is also asked to provide an optional API budget dollar amount. If they enter nothing, there is no monetary limit, but if they define a budget then the AI will be told to shut down gracefully once it has come within 1 cent of its limit, and to shut down immediately once it has exceeded its limit. The AI is always aware of its initial and remaining budget.

### Changes
- Centralized chat completion API calls through `ApiManager` class in `api_manager.py`
- Removed openai imports from other files to prevent accidental API calls
- Updated `AIConfig` class to include an optional `api_budget` parameter
- Modified `chat_with_ai` function to include `api_budget` and display remaining budget to the AI
- Updated `browse.py`, `call_ai_function.py`, `chat.py`, and `main.py` to use the centralized `ApiManager`
- Added input for API budget to `load_variables` and `construct_prompt` functions in `main.py`

### Documentation
There are Docstrings and in-code comments documenting things where necessary

### Test Plan
I have tested this feature in the following ways:
- Various different prompts were used, assigning Auto-GPT to various different tasks, and it did not encounter any more errors than before the change
- I put an empty string for the budget, and the AI proceeded as if it had infinite budget
- I put a dollar amount for the budget, and the AI proceeded as if it had a limited budget, hurrying itself up as the budget depleted
- The AI shuts itself down once the budget is exceeded
- The total running cost is being printed to the console for the user to see upon every chat completion, and the amount makes sense for the models used

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Review by richbeales — DISMISSED** [review]

**Comment by Vwing** [comment]
> The change scope is too big and it needs to be split. So I suggest just to extract the api_manager from the code first, preserving the current behavior. And then change the behavior in a separate PR.

Sure! I could make a new PR with just the first two commits. Those only deal with adding the api_manager, routing chat completions through it, and printing the total running cost to the console.

[e168ab0](https://github.com/Torantulino/Auto-GPT/pull/762/commits/e168ab014bd563a13bfa008c67d9f9d51e9da34b)
[8492a7f](https://github.com/Torantulino/Auto-GPT/pull/762/commits/8492a7f8be9f633a5009a43e8f5550be00463c97)

**Comment by nponeccop** [comment]
@Torantulino we need you to approve the plan

**Comment by Vwing** [comment]
@nponeccop @Torantulino Shall I implement the plan detailed above? In my new PR, I can also update the code with the latest changes to master.

**Review by nponeccop — DISMISSED** [review]

File: autogpt/api_manager.py
Comment (rob-luke) [inline autogpt/api_manager.py]:
```suggestion
# Define the cost per thousand tokens for each model
```

**Comment by rob-luke** [comment]
I am so pleased someone created this PR, thanks @Vwing.
This does not track the cost of the embedding calls. While this could be it be added in a future PR, it may confuse end users when their bill is larger than they expect. I am happy to contribute this extra feature if you want, just let me know.

**Comment by Vwing** [comment]
Sure, @rob-luke. That would be much appreciated!

**Comment by Vwing** [comment]
@rob-luke Heads up, I merged in master to resolve the conflicts.

**Comment by Pwuts** [comment]
Almost there, please fix the linting errors 🥁

**Comment by Vwing** [comment]
@Pwuts done!

**Comment by rob-luke** [comment]
I dont want to block this from getting merged, so I will just create a follow up PR once this is merged

**Comment by rob-luke** [comment]
I added the feature and opened a PR in to this branch here: https://github.com/Vwing/Auto-GPT/pull/1
But if you merge this branch as is, then I will just open a fresh PR in to master of this repo later.
Thanks again for the great package and for this PR.

**Comment by Vwing** [comment]
I merged in your changes, @rob-luke, they look good! Just make sure next time to run the linter 
image

**Comment by Vwing** [comment]
@BillSchumacher I don't know if you've started reviewing this PR yet, but just to let you know I merged in @rob-luke's changes so it tracks embedding costs.

**Comment by rob-luke** [comment]
Thanks Vwing, and sorry for missing the linting. Thanks for fixing my errors

**Comment by ntindle** [comment]
Can you update from the latest? theres some changes to the pipelines that need to run @Vwing

**Review by ntindle — CHANGES_REQUESTED** [review]
update to latest

**Comment by Vwing** [comment]
Yep! I'm on it.

**Comment by ntindle** [comment]
Give it a bit, you’ll have to do it twice if not

**Comment by Vwing** [comment]
Okay, I screwed up and merged too soon 😅

I'm going to try to clean up my history. I'll be a moment...

**Review by ntindle — DISMISSED** [review]

**Comment by ntindle** [comment]
Heads up this may need scheduled in

**Comment by Vwing** [comment]
Oops, some tests failed. I can fix that...

**Comment by Vwing** [comment]
Hi @ntindle, I notice that the test is failing because it's using the "gpt2" model, which isn't an option for OpenAI chat completions. I'm wondering how we should proceed in this case. Should I check for an invalid model and fail quietly, or do you think it's better to update the test to use a valid model? Looking forward to your guidance, thanks!

**Comment by ntindle** [comment]
Update it and wrap it with @requires_API_key("Open_AI_KEY")

**Comment by Vwing** [comment]
@ntindle Okay, done and done. Fingers crossed it passes the tests this time. I don't have an azure openai key so I can't run all of the tests locally.

File: autogpt/api_manager.py
Comment (ntindle) [inline autogpt/api_manager.py]: Have this read from the config debug setting

File: autogpt/api_manager.py
Comment (ntindle) [inline autogpt/api_manager.py]: Can we extract this to a models.json or a modelsinfo.py. It's duplicated throughout the codebsse and it would be nice to just say use gpt3 and pass that as an object

File: autogpt/api_manager.py
Comment (ntindle) [inline autogpt/api_manager.py]: Add a debug param

File: autogpt/api_manager.py
Comment (ntindle) [inline autogpt/api_manager.py:47]: Debug log the response

File: autogpt/chat.py
Comment (ntindle) [inline autogpt/chat.py:140]: Debug log that this is occurring

**Review by ntindle — CHANGES_REQUESTED** [review]
I'd love to see full test coverage of the api mamager

**Comment by Vwing** [comment]
I have made the requested changes.

One note, however: now that the 'print_total_cost' is confined to debug mode, the user will no longer have awareness of their running cost during normal use, which was one of the goals outlined in issue #6. Perhaps this feature can be added in a future pull request.

**Comment by ntindle** [comment]
You’re right on that comment. May be worth having it do both in the future

**Review by ntindle — APPROVED** [review]

**Pull request merged**
